### PR TITLE
Add flake.nix and Cargo.nix

### PR DESCRIPTION
**Pull request by AkechiShiro**
Add for easier use of the tool under Nix/NixOS

### CODE REVIEW

**Comment by AkechiShiro** [comment]
Hello @orangecms :wave: , I just wanted to gently ping you regarding this PR. 

Whenever you have a moment, your insight would be greatly appreciated!

**Comment by orangecms** [comment]
Oh hey sorry, I wasn't subscribed to notifications here... gotta actively do that for each repo, IIUC; I have so many and forget to do that. :sweat_smile: 

I'm glad to see that you looked into this tool after all and tried it out! :)
Is it working fine for you, or do you need anything? I had a few other things I just did, not sure when I'll get back to this.

Re the changes: I'd prefer to keep external packaging and build utilities outside of the project, just stick with the native Rust tools. We already have a lockfile, so another copy would create a maintenance burden when making changes. I think you can just keep it in a fork a regenerate ad-hoc, right?

**Comment by AkechiShiro** [comment]
It's totally okay, no worries !

I need more time unfortunately to fully try the tool, I haven't compiled a kernel with full access to `/dev/mem`, I haven't found the time to do that, just yet but I want to dive more into it.

I saw some of your videos on YouTube about firmware attacks (I only saw part 1 so far, and it's pretty good stuff, thanks for the content), I'm not sure if you make use of `ems` in them.

I can totally keep it in a fork if you wish so, thing is it wasn't too much of a hassle to create, I used https://github.com/cargo2nix/cargo2nix 
```
# Use nix to get cargo2nix & rust toolchain on your path
nix develop github:cargo2nix/cargo2nix#bootstrap

# In directory with Cargo.toml & Cargo.lock files (cargo generate-lockfile)
cargo2nix

# Or skip the shell and run it directly
nix run github:cargo2nix/cargo2nix

# You'll need this in version control
git add Cargo.nix
```

I believe a GitHub actions could probably do the trick and maintain it based on the cargo lockfile, but I think it's okay to leave it in a fork for now, I just wanted to make it easier for the Nix community to run the ems tool.

If this PR is accepted, running and installing `ems` dependencies, can be done like so 
```
nix run github:platform-system-interface/ems
```
But with the rust toolchain, it is also not too complicated, so I'm not sure if it's worth having a GitHub actions and implement all this at the current stage of this project, this might be too early.

**Comment by orangecms** [comment]
The development of `ems` started in the third video, then just called `findme`. :)
I had improved it a bit more over 37C3 so that it has some basic idea of some of the EFI structs.

The code could be taken in part to run directly as an EFI app, so you could use the x86 EFI target (Rust supports that now :)) and run it in QEMU directly in combo with OVMF.

Roughly, this should do:

```
qemu-system-x86_64 -kernel ems-for-uefi -bios OVMF.fd -nographic
```

**Comment by orangecms** [comment]
If I understand correctly, Nix users could just clone the repo and then `cargo2nix && nix run .`?

That would always work, is just simple to run, so there is no need to have an extra copy of the generated files in this repo. I feel like that is something that Nix users would find in the Nix manual. WDYT?

**Comment by AkechiShiro** [comment]
I was at 37C3 during a meetup where you presented very quickly this tool.

Ah I see, so I can run this tool inside QEMU as an EFI app, okay, that's good to know, I didn't know it was an EFI app.

With the `flake.nix` file in this PR, nix users will only have to run : `nix run github:platform-system-interface/ems` and it will get all the dependencies and compile them for the user.

And yes Nix users, could clone the repo, run `cargo2nix` but this wouldn't be enough, to consume the `Cargo.nix` file generated : 
image


They require `flake.nix` file, I hacked one for the repo, it was very simple.

**Comment by AkechiShiro** [comment]
I believe that it would make it easier, if there was a `flake.nix` file in the repo however the very good/proper solution for Nix users, would be to package `ems` in Nixpkgs (the official repository of Nixpkgs) and PR on your README, instructions to install it that way.

You can see flakes as *kind* of workaround that can be used to install `non-officially` maintained packages in Nix.

I'm sorry, I'm not sure if I'm being clear enough in my comments/explanation, Nix may be something you're not familiar with.

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  outputs = inputs: with inputs;
+    flake-utils.lib.eachDefaultSystem (system:
+      let
+        pkgs = import nixpkgs {
+          inherit system;
+          overlays = [cargo2nix.overlays.default];
+        };
+
+        rustPkgs = pkgs.rustBuilder.makePackageSet {
+          rustVersion = "1.70.0";
```
Comment (AkechiShiro) [inline flake.nix:17]: Here, I might need to update the rust version to 1.75.0 as well.

I think it's better to leave this into a fork and probably update the README, to point to it for Nix users.

File: flake.nix [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{
+  inputs = {
+    cargo2nix.url = "github:cargo2nix/cargo2nix/release-0.11.0";
```
Comment (AkechiShiro) [inline flake.nix:3]: This release version might also need to be updated in the future.

**Comment by orangecms** [comment]
> I was at 37C3 during a meetup where you presented very quickly this tool.

yea I remember :)

> Ah I see, so I can run this tool inside QEMU as an EFI app, okay, that's good to know, I didn't know it was an EFI app.

Not yet, I created an issue for creating that.

> With the `flake.nix` file in this PR, nix users will only have to run : `nix run github:platform-system-interface/ems` and it will get all the dependencies and compile them for the user.
> 
> And yes Nix users, could clone the repo, run `cargo2nix` but this wouldn't be enough, to consume the `Cargo.nix` file generated : 
> image
> 
> 
> They require `flake.nix` file, I hacked one for the repo, it was very simple.

Yes, that would be a simple command, maybe with a switch to declare the Rust version. A shell script would do. Something like `nix-cargo-run github.com/...`. So that would do all in one shot:
- clone the repo
- generate the Nix files (flake and cargo2nix)
- run it

I am very familiar with Nix(OS) from work. So I have a created a bunch of shell scripta to simplify its use, e.g. a `nixup` script to upgrade the system. ^^

**Comment by AkechiShiro** [comment]
Oh sorry, I wasn't sure if you were familiar with Nix/NixOS, I will be more careful in properly asking before starting to make assumption, I apologize.

You are correct about the shell script, it''s a way better idea, that would not pollute the repository for users that use Nix.

**Comment by AkechiShiro** [comment]
Then probably this PR can be superseded by another PR, that implement the shell script?

Or should I just continue on this PR and implement a shell script that does what we discussed ?

**Comment by orangecms** [comment]
> Then probably this PR can be superseded by another PR, that implement the shell script?
> 
> Or should I just continue on this PR and implement a shell script that does what we discussed ?

I think that script would best live on its own, so it can be used for other repos as well and maintained on its own. :)
Maybe Nix would even take it upstream? It is really a convenience tool specifically for Nix. :-)

**Comment by AkechiShiro** [comment]
You are most likely right, I'm not sure if such a script already exists, I should better do my research on this side, this PR can be closed, I suppose ?

**Pull request closed without merging**